### PR TITLE
New class for state handling; set cookie for implicit nonce

### PR DESCRIPTION
**Pull request by joshcanhelp**
- Refactored state and nonce handling to extend a `WP_Auth0_Random_Storage` class. 
- Set a `nonce` cookie for implicit logins during Lock init
- Set a `nonce` cookie for implicit auto-logins before redirect
- Renamed methods to conform to WP code standards

### CODE REVIEW

File: lib/WP_Auth0_Nonce_Handler.php [hunk 1]
```diff
@@ -1,134 +1,24 @@
 <?php
 
-final class WP_Auth0_Nonce_Handler {
```
Comment (joshcanhelp) [inline lib/WP_Auth0_Nonce_Handler.php]: Move to parent class

File: lib/WP_Auth0_Random_Storage.php [hunk 2]
```diff
@@ -0,0 +1,121 @@
+<?php
+
+class WP_Auth0_Random_Storage {
```
Comment (joshcanhelp) [inline lib/WP_Auth0_Random_Storage.php]: This was previously the `WP_Auth0_Nonce_Handler`. I added a more generic cookie name and removed the `verify` method so extending classes can implement it themselves.

File: lib/WP_Auth0_LoginManager.php [hunk 3]
```diff
@@ -146,15 +135,15 @@ public function init_auth0() {
 
 		// Catch any incoming errors and stop the login process.
 		// See https://auth0.com/docs/libraries/error-messages for more info.
-		if ( $this->query_vars( 'error' ) || $this->query_vars( 'error_description' ) ) {
-			$error_msg  = sanitize_text_field( rawurldecode( $_REQUEST[ 'error_description' ] ) );
-			$error_code = sanitize_text_field( rawurldecode( $_REQUEST[ 'error' ] ) );
+		if ( ! empty( $_REQUEST['error'] ) || ! empty( $_REQUEST['error_description'] ) ) {
```
Comment (joshcanhelp) [inline lib/WP_Auth0_LoginManager.php]: `query_vars` was causing problems in the state retrieval so I changed that here as well.

File: lib/WP_Auth0_LoginManager.php [hunk 4]
```diff
@@ -228,8 +217,7 @@ public function redirect_login() {
 
 			// Look for clues as to what went wrong.
 			$e_message = ! empty( $data->error_description ) ? $data->error_description : __( 'Unknown error', 'wp-auth0' );
-			$e_code    = ! empty( $data->error ) ? $data->error : $exchange_resp_code;
-			throw new WP_Auth0_LoginFlowValidationException( $e_message, $e_code );
+			throw new WP_Auth0_LoginFlowValidationException( $e_message, $exchange_resp_code );
```
Comment (joshcanhelp) [inline lib/WP_Auth0_LoginManager.php]: Didn't like the string error code ... found during debugging

File: lib/WP_Auth0_LoginManager.php [hunk 5]
```diff
@@ -608,30 +595,30 @@ public static function get_userinfo_scope( $context = '' ) {
 	 * @return array
 	 */
 	public static function get_authorize_params( $connection = null, $redirect_to = null ) {
```
Comment (joshcanhelp) [inline lib/WP_Auth0_LoginManager.php]: Mainly whitespace changes in this method

File: lib/WP_Auth0_LoginManager.php [hunk 6]
```diff
@@ -326,7 +314,7 @@ public function implicit_login() {
 
 		// Validate the nonce if one was included in the request if using auto-login.
 		$nonce = isset( $decoded_token->nonce ) ? $decoded_token->nonce : null;
-		if ( ! WP_Auth0_Nonce_Handler::getInstance()->validate( $nonce ) ) {
+		if ( ! WP_Auth0_Nonce_Handler::get_instance()->validate( $nonce ) ) {
```
[the remark this answers is not in the document]
Comment (joshcanhelp) [reply lib/WP_Auth0_LoginManager.php]: For new methods, I'm following the WP standard, which is snake-case. There was no naming convention, really, up until now.

File: lib/WP_Auth0_LoginManager.php [hunk 7]
```diff
@@ -103,31 +87,36 @@ public function init() {
...
 
-		$connection = apply_filters( 'auth0_get_auto_login_connection', $this->a0_options->get( 'auto_login_method' ) );
+		$connection  = apply_filters( 'auth0_get_auto_login_connection', $this->a0_options->get( 'auto_login_method' ) );
 		$auth_params = self::get_authorize_params( $connection );
 
 		$auth_url = 'https://' . $this->a0_options->get( 'domain' ) . '/authorize';
 		$auth_url = add_query_arg( array_map( 'rawurlencode', $auth_params ), $auth_url );
 
-		setcookie( WPA0_STATE_COOKIE_NAME, $auth_params['state'], time() + WP_Auth0_Nonce_Handler::COOKIE_EXPIRES, '/' );
+		WP_Auth0_State_Handler::get_instance()->setStateCookie( $auth_params['state'] );
```
[the remark this answers is not in the document]
Comment (joshcanhelp) [reply lib/WP_Auth0_LoginManager.php]: Correct. That array comes from the `get_authorize_params()` method in this same class and is always generated.

File: lib/WP_Auth0_Nonce_Handler.php [hunk 8]
```diff
@@ -1,134 +1,40 @@
...
-   * @return string
-   */
-  public function generateNonce( $bytes = 32 ) {
-    $nonce_bytes = function_exists( 'random_bytes' ) ? random_bytes( $bytes ) : openssl_random_pseudo_bytes( $bytes );
-    return bin2hex( $nonce_bytes );
-  }
-}
\ No newline at end of file
+/**
+ * Contains WP_Auth0_Nonce_Handler.
```
[the remark this answers is not in the document]
Comment (joshcanhelp) [reply lib/WP_Auth0_Nonce_Handler.php]: It's a file and a class comment. Duplicative but part of the code quality scan.

File: lib/WP_Auth0_Random_Storage.php [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+	}
+
+	/**
+	 * Reset/delete a cookie.
+	 *
+	 * @param null|string $cookie_name - name of the cookie to reset.
+	 *
+	 * @return bool
+	 */
+	public function reset( $cookie_name = null ) {
```
[the remark this answers is not in the document]
Comment (joshcanhelp) [reply lib/WP_Auth0_Random_Storage.php]: `null` means clear the uniqid cookie. Passing a value means clear a different cookie.

File: lib/WP_Auth0_State_Handler.php [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+	 * @var string
+	 */
+	const STATE_COOKIE_NAME = 'auth0_state';
+
+	/**
+	 * Singleton class instance
+	 *
+	 * @var WP_Auth0_State_Handler|null
+	 */
+	protected static $_instance = null;
```
[the remark this answers is not in the document]
Comment (joshcanhelp) [reply lib/WP_Auth0_State_Handler.php]: See above ☝️

File: lib/WP_Auth0_Random_Storage.php [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+	 * @var integer
+	 */
+	const COOKIE_EXPIRES = HOUR_IN_SECONDS;
+
+	/**
+	 * Singleton class instance.
+	 *
+	 * @var WP_Auth0_Random_Storage|null
+	 */
+	protected static $_instance = null;
```
[the remark this answers is not in the document]
Comment (joshcanhelp) [reply lib/WP_Auth0_Random_Storage.php]: See above ☝️

File: lib/WP_Auth0_State_Handler.php [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+	 * Check if the stored nonce matches the URL value.
+	 *
+	 * @return bool
+	 */
+	public function validate() {
+		$valid = isset( $_COOKIE[ self::STATE_COOKIE_NAME ] )
+			? $_COOKIE[ self::STATE_COOKIE_NAME ] === $this->getState()
+			: false;
+		$this->reset();
+		$this->reset( static::STATE_COOKIE_NAME );
```
[the remark this answers is not in the document]
Comment (joshcanhelp) [reply lib/WP_Auth0_State_Handler.php]: Refactoring this a bit based on your feedback ...

File: lib/WP_Auth0_Random_Storage.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (joshcanhelp) [reply lib/WP_Auth0_Random_Storage.php]: Refactoring this a bit based on your feedback ...

File: lib/WP_Auth0_State_Handler.php [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+	}
+
+	/**
+	 * Set the state cookie value.
+	 *
+	 * @param string $value - state value to store, same as what is sent to Auth0.
+	 *
+	 * @return bool
+	 */
+	public function setStateCookie( $value ) {
```
[the remark this answers is not in the document]
Comment (joshcanhelp) [reply lib/WP_Auth0_State_Handler.php]: Refactoring this a bit based on your feedback ... also bad method casing here

File: lib/WP_Auth0_State_Handler.php [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+		$_COOKIE[ self::STATE_COOKIE_NAME ] = $value;
+		return setcookie( self::STATE_COOKIE_NAME, $value, $this->get_cookie_exp(), '/' );
+	}
+
+	/**
+	 * Get the returned state from a URL parameter.
+	 *
+	 * @return string
+	 */
+	public function getState() {
```
[the remark this answers is not in the document]
Comment (joshcanhelp) [reply lib/WP_Auth0_State_Handler.php]: Bad method casing here :( correcting. 

I'll remove this, it's only used in `validate` above, over-abstracted looking at it a second time.

File: lib/WP_Auth0_State_Handler.php [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+ * @package WP-Auth0
+ */
+
+/**
+ * Class WP_Auth0_State_Handler for handling state storage and verification.
+ */
+final class WP_Auth0_State_Handler extends WP_Auth0_Random_Storage {
+
+	/**
+	 * Cookie name used to store nonce
```
[the remark this answers is not in the document]
Comment (joshcanhelp) [reply lib/WP_Auth0_State_Handler.php]: `nonce` is correct in this case. This nonce is added to a state object

File: lib/WP_Auth0_Nonce_Handler.php [hunk 16]
```diff
@@ -1,134 +1,40 @@
...
+ *
+ */
+final class WP_Auth0_Nonce_Handler extends WP_Auth0_Random_Storage {
+
+	/**
+	 * Cookie name used to store nonce
+	 *
+	 * @var string
+	 */
+	const UNIQID_COOKIE_NAME = 'auth0_nonce_uniqid';
```
[the remark this answers is not in the document]
Comment (joshcanhelp) [reply lib/WP_Auth0_Nonce_Handler.php]: Synonymous as used. The uniqid is used as a nonce in the Nonce class.

File: lib/WP_Auth0_State_Handler.php [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+	 * @var string
+	 */
+	const UNIQID_COOKIE_NAME = 'auth0_state_uniqid';
+
+	/**
+	 * Cookie name used to store nonce
+	 *
+	 * @var string
+	 */
+	const STATE_COOKIE_NAME = 'auth0_state';
```
[the remark this answers is not in the document]
Comment (joshcanhelp) [reply lib/WP_Auth0_State_Handler.php]: State is state, nonce is the salt in state.

File: lib/WP_Auth0_Nonce_Handler.php [hunk 18]
```diff
@@ -1,134 +1,40 @@
...
+
+	/**
+	 * Singleton class instance
+	 *
+	 * @var WP_Auth0_Nonce_Handler|null
+	 */
+	protected static $_instance = null;
```
[the remark this answers is not in the document]
Comment (lbalmaceda) [reply lib/WP_Auth0_Nonce_Handler.php]: you already have a `getInstance` method (the singleton getter) on the super class. Use that from the child ones.

[the remark this answers is not in the document]
Comment (joshcanhelp) [reply lib/WP_Auth0_Nonce_Handler.php]: Yes, I do have that and do use that but the child classes need to have an `_instance` property for that to work.

File: lib/WP_Auth0_LoginManager.php [hunk 19]
```diff
@@ -146,15 +135,16 @@ public function init_auth0() {
 
 		// Catch any incoming errors and stop the login process.
 		// See https://auth0.com/docs/libraries/error-messages for more info.
-		if ( $this->query_vars( 'error' ) || $this->query_vars( 'error_description' ) ) {
-			$error_msg  = sanitize_text_field( rawurldecode( $_REQUEST[ 'error_description' ] ) );
-			$error_code = sanitize_text_field( rawurldecode( $_REQUEST[ 'error' ] ) );
+		if ( ! empty( $_REQUEST['error'] ) || ! empty( $_REQUEST['error_description'] ) ) {
```
Comment (lbalmaceda) [inline lib/WP_Auth0_LoginManager.php]: would an `error_description` by itself be valid? I guess only when `error` is present. So 2nd clause could be removed.

File: lib/WP_Auth0_LoginManager.php [hunk 20]
```diff
@@ -146,15 +135,16 @@ public function init_auth0() {
...
+			$error_msg  = sanitize_text_field( rawurldecode( $_REQUEST['error_description'] ) );
+			$error_code = sanitize_text_field( rawurldecode( $_REQUEST['error'] ) );
 			$this->die_on_login( $error_msg, $error_code );
 		}
 
 		// Check for valid state nonce, set in WP_Auth0_Lock10_Options::get_state_obj().
 		// See https://auth0.com/docs/protocols/oauth2/oauth-state for more info.
-		if ( ! $this->validate_state() ) {
+		$state_returned = isset( $_REQUEST['state'] ) ? rawurldecode( $_REQUEST['state'] ) : null;
+		if ( ! $state_returned || ! WP_Auth0_State_Handler::get_instance()->validate( $state_returned ) ) {
```
Comment (lbalmaceda) [inline lib/WP_Auth0_LoginManager.php]: `! $state_returned ` is already covered inside the verify method

File: lib/WP_Auth0_Nonce_Handler.php [hunk 21]
```diff
@@ -1,134 +1,180 @@
...
+	/**
+	 * Start-up process to make sure we have something stored.
+	 */
+	private function init() {
+		if ( isset( $_COOKIE[ static::UNIQID_COOKIE_NAME ] ) ) {
+			// Have a cookie, don't want to generate a new one.
+			$this->_uniqid = $_COOKIE[ static::UNIQID_COOKIE_NAME ];
+		} else {
+			// No cookie, need to create one.
+			$this->_uniqid = $this->generate_nonce();
```
Comment (lbalmaceda) [inline lib/WP_Auth0_Nonce_Handler.php]: don't you need to set the generated value to `$_COOKIE`?

File: lib/WP_Auth0_Nonce_Handler.php [hunk 22]
```diff
@@ -1,134 +1,180 @@
...
+	public function get_cookie_exp() {
+		return time() + self::COOKIE_EXPIRES;
+	}
+
+	/**
+	 * Set the random storage cookie.
+	 *
+	 * @return bool
+	 */
+	public function set_cookie() {
```
Comment (lbalmaceda) [inline lib/WP_Auth0_Nonce_Handler.php]: shouldn't this be called internally after generating a new nonce? (private). It doesn't receive any param so I don't see the point of it being public at all

**Pull request merged**